src/contracts/atlas/GasAccounting.sol

```diff
@@ -201,107 +224,190 @@ abstract contract GasAccounting is SafetyLocks {
+                // NOTE: We assume the bundler is fully on the hook for the gas, so we ignore claims and writeoffs
+                // and focus exclusively on deposits and withdrawals.
+                // The "+ netAtlasGasSurcharge" forces Bundler to pay the surcharge
+                revert InsufficientTotalBalance(_withdrawals + netAtlasGasSurcharge - _deposits);
+            }
+            claimsPaidToBundler = _deposits - _withdrawals;
+        } else if (_winningSolver == ctx.bundler) {
+            // CASE: The winning solver is also the bundler
+            if (_deposits < _withdrawals + netAtlasGasSurcharge) {
+                // CASE: in deficit, subtract from bonded balance
+                uint256 amountOwed = _withdrawals + netAtlasGasSurcharge - _deposits;
+                uint256 deficit = _assign(_winningSolver, amountOwed, true);
+                if (deficit > 0) {
+                    revert InsufficientTotalBalance(deficit); // Revert if insufficient bonded balance.
+                }
+                claimsPaidToBundler = 0; // Bundler-Solver is in deficit, no need to pay them.
+            } else {
+                // CASE: in surplus, add to bonded balance by crediting the bundler at bottom of function
+                claimsPaidToBundler = _deposits - _withdrawals - netAtlasGasSurcharge;
+            }
+        } else if (_writeoffs + _deposits < _claims + _withdrawals) {
+            // CASE: Not a special bundler, solver successful, balance in deficit.
+            // NOTE _claims and _writeoffs already have the Gas Surcharges factored in
+            uint256 amountOwed = _claims + _withdrawals - _writeoffs - _deposits;
+            uint256 deficit = _assign(_winningSolver, amountOwed, true);
             if (deficit > 0) {
-                revert InsufficientTotalBalance(deficit);
+                // CASE: Solver's bonded balance isn't enough to cover the amount owed, and the
+                // winning solver is unrelated to the bundler. The Bundler is not the Solver.
+                if (deficit > _claims - _writeoffs - netAtlasGasSurcharge) {
+                    // CASE: The deficit is too large to writeoff by having the Bundler absorb the cost
+                    revert InsufficientTotalBalance(deficit);
+                } else {
+                    // CASE: We can writeoff the deficit (Bundler has already paid the gas anyway)
+                    // TODO: Ensure incentive compatibility
+                    _writeoffs += deficit;
+                }
             }
+            claimsPaidToBundler = _claims - _writeoffs - netAtlasGasSurcharge;
         } else {
             // CASE: in surplus, add to bonded balance
-            // TODO: make sure this works w/ the surcharge 10%
-            uint256 amountCredited = _deposits - _claims - _withdrawals;
-            _credit(winningSolver, amountCredited);
+            uint256 amountCredited = _deposits + _writeoffs - _claims - _withdrawals;
```

Comment:
Maybe I'm confusing something here. But is there something off with how we account for `deposits` and `withdrawals`?

In `_setAccountingLock()`, any ETH the _bundler_ sent is added to `deposit`, but then here it looks like `deposits` positively affect the _solver_ as any excess is credited back to them.

E.g. msg.value = 1 ETH, userOpValue = 0, writeoffs = 0, solver doesn't borrow anything, and `_claims < _deposits`. Then, wouldn't we hit this `else` block and add `1 ETH - claims` to the solver's bonded balance via `_credit()`?

We would also have a positive `claimsPaidToBundler` and `msg.value (1 ETH) > claimsPaidToBundler` so we then also transfer `claims - netAtlasGasSurcharge` back to the bundler.

Is this intentional - that any ETH the bundler sends in metacall that is "left over" is sent to the winning solver and not back to the bundler?

Reply:
From Discord discussion: This is intentional. Bundlers also see the `userOp`, the `solverOps`, and the `dAppOp` before choosing to attach a `msg.value` to their `metacall` tx, so they have access to all the information needed to make the decision.

We should also document that the only reason to send ETH with the `metacall` tx is to subsidize the winning solver's gas cost repayment, and bundlers should be aware that they may not receive all the ETH sent with the metacall afterwards.